### PR TITLE
executor: fix last_insert_id in auto_random mode

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb/issues/15140.
- Last insert id shouldn't be 0 when inserting with auto_random.
- 0 is not replaced with an auto allocated value.
- Error message of auto_random is confusing.

### What is changed and how it works?
- Set last insert id after inserting 0.
- Consider `NO_AUTO_VALUE_ON_ZERO` and fill auto value with inserting 0.
- Refine error message of auto_random.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
```
mysql> insert ar value(0, 'a'), (0, 'b');
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> select * from ar;
+-----------+------+
| id        | name |
+-----------+------+
| 671178642 | a    |
| 671178643 | b    |
+-----------+------+
2 rows in set (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|        671178642 |
+------------------+
1 row in set (0.01 sec)
```

```
mysql> create table ar(id int key auto_random(32));
ERROR 8216 (HY000): Invalid auto random: Bits of column `id` is 32, but auto_random bits is 32. Max allowed auto_random bits for column `id` is 31
mysql> create table ar(id int key auto_random(31));
Query OK, 0 rows affected (0.00 sec)
```

Code changes

N/A

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Fix the bug that `last_insert_id` is wrong when inserting with `auto_random`; Inserting 0 with `auto_random` column is not replaced with an allocated value.

### CODE REVIEW

File: executor/insert_common.go [hunk 1]
```diff
@@ -847,30 +847,50 @@ func (e *InsertValues) adjustAutoRandomDatum(ctx context.Context, d types.Datum,
 		return d, nil
 	}
 
-	if !hasValue || d.IsNull() {
-		_, err := e.ctx.Txn(true)
```
Comment (AilinKid) [inline executor/insert_common.go]: better keep this, cause `allocAutoRandomID` will use `txn.startTs` to shard.

Comment (djshow832) [reply executor/insert_common.go]: Txn is moved to line 876.

File: executor/insert_test.go [hunk 2]
```diff
@@ -958,3 +959,166 @@ func (s *testSuite3) TestAutoIDIncrementAndOffset(c *C) {
...
+
+	tk.MustExec(`insert into ar(name) values ('a')`)
+	rs = tk.MustQuery(`select id from ar`)
+	c.Assert(len(rs.Rows()), Equals, 1)
+	firstValue, err = strconv.Atoi(rs.Rows()[0][0].(string))
+	c.Assert(err, IsNil)
+	c.Assert(firstValue, Greater, 0)
+	tk.MustQuery(`select last_insert_id()`).Check(testkit.Rows(fmt.Sprintf("%d", firstValue)))
+
+	tk.MustExec(`drop table ar`)
```
Comment (AilinKid) [inline executor/insert_test.go]: how about do rebase first，then check the `last_insert_id`.
eg：insert into ar(id) value(100),(NULL),  the `last_insert_id` should be 101.

File: executor/insert_common.go (hunk 1 above)
Comment (AilinKid) [reply executor/insert_common.go]: nice

File: executor/insert_test.go (hunk 2 above)
Comment (djshow832) [reply executor/insert_test.go]: Situations are too many, we should add many test cases in the future. So I ignore some cases for now.

Comment (AilinKid) [reply executor/insert_test.go]: fine.

**Pull request merged**